ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java

```diff
@@ -306,16 +310,18 @@ public synchronized Container getContainerById(int accountId, int containerId) t
    * @throws SQLException
    */
   public synchronized Dataset addDatasetVersions(int accountId, int containerId, String accountName,
-      String containerName, String datasetName, long version, long expirationTimeMs) throws SQLException {
+      String containerName, String datasetName, String version, long expirationTimeMs) throws SQLException {
     try {
       long startTimeMs = System.currentTimeMillis();
       dataAccessor.getDatabaseConnection(true);
       PreparedStatement insertDatasetVersionStatement =
           dataAccessor.getPreparedStatement(insertDatasetVersionSql, true);
       Dataset dataset =
           getDatasetMetadata(accountId, containerId, accountName, containerName, datasetName, expirationTimeMs);
-      executeAddDatasetVersionStatement(insertDatasetVersionStatement, accountId, containerId, datasetName, version,
-          expirationTimeMs, dataset);
+      //TODO: if version == null, add the latest version from db + 1.
+      long versionNumber = getVersionBasedOnSchema(version, dataset.getVersionSchema());
+      executeAddDatasetVersionStatement(insertDatasetVersionStatement, accountId, containerId, datasetName,
+          versionNumber, expirationTimeMs, dataset);
```

Comment:
You are right, after put the getDataset before the datasetVersion query, it can help detect it. And I also have an pr which include the resultSet.next() == null check, so it will throw the error messages.